Typos (#2126)
javascript-tutorial/en.javascript.info: 6-data-storage/03-indexeddb/article.md

**IndexedDB has a built-in mechanism of "schema versioning", absent in server-side databases.**

> Unlike server-side databases, IndexedDB is client-side, the data is stored in the browser, so we, developers, don't have full-time access to it. So, when we have published a new version of our app, and the user visits our webpage, we may need to update the database.

Reviewer: ```suggestion
Unlike server-side databases, IndexedDB is client-side - the data is stored in the browser. So we, developers, don't have full-time access to it. Due to that, we can only change/update the database when the user visits our webpage.
```

```warn header="We can't open an older version of the database"
If the current user database has a higher version than in the `open` call, e.g. the existing DB version is `3`, and we try to `open(...2)`, then that's an error, `openRequest.onerror` triggers.

That's rare, but such a thing may happen when a visitor loads outdated JavaScript code, e.g. from a proxy cache. So the code is old, but his database is new.

> To protect from errors, we should check `db.version` and suggest he reloads the page. Use proper HTTP caching headers to avoid loading the old code, so that you'll never have such problems.

Reviewer: ```suggestion
To protect from errors, we should check `db.version` and suggest a page reload. Use proper HTTP caching headers to avoid loading the old code, so that you'll never have such problems.
```

Reviewer: Actually, I am not quite sure if your usage is grammatically correct. Something just doesn't feel right. I am not sure, so you could be right.

Author: Yes, you are right, it's a subjunctive. Glad you caught it.

The problem is that a database is shared between two tabs, as it's the same site, same origin. And it can't be both version `1` and `2`. To perform the update to version `2`, all connections to version 1 must be closed, including the one in the first tab.

> In order to organize that, the `versionchange` event triggers on the "outdated" database object. We should listen for it and close the old database connection (and probably suggest the visitor reloads the page, to load the updated code).

Reviewer: ```suggestion
In order to organize that, the `versionchange` event triggers on the "outdated" database object. We should listen for it and close the old database connection (and probably suggest the visitor to reload the page, to load the updated code).
```

Reviewer: OR

```suggestion
In order to organize that, the `versionchange` event triggers on the "outdated" database object. We should listen for it and close the old database connection (and probably suggest a page reload, to load the updated code).
```

IndexedDB uses the [standard serialization algorithm](https://www.w3.org/TR/html53/infrastructure.html#section-structuredserializeforstorage) to clone-and-store an object. It's like `JSON.stringify`, but more powerful, capable of storing much more datatypes.

An example of an object that can't be stored: an object with circular references. Such objects are not serializable. `JSON.stringify` also fails for such objects.

**There must be a unique `key` for every value in the store.**     

> A key must be one of the these types: number, date, string, binary, or array. It's a unique identifier: we can search/remove/update values by the key.

Reviewer: ```suggestion
A key must be one of the these types - number, date, string, binary, or array. It's a unique identifier, so we can search/remove/update values by the key.
```

Transactions auto-commit principle has an important side effect. We can't insert an async operation like `fetch`, `setTimeout` in the middle of transaction. IndexedDB will not keep the transaction waiting till these are done.

> In the code below `request2` in line `(*)` fails, because the transaction is already committed, and can't make any request in it:

Reviewer: ```suggestion
In the code below, `request2` in line `(*)` fails, because the transaction is already committed, and can't make any request in it:
```
